Push ServeSession to the start (#3)
Dekkonot/rojo: ARCHITECTURE.md

## Overarching

> Rojo is divide into several components, each layering on top of each other provide Rojo's functionality.

Reviewer: ```suggestion
Rojo is divided into several components, each layering on top of each other provide Rojo's functionality.
```
Fix tense

> At the core of Rojo, is `ServeSession`. As the name implies, it contains all of the components to keep a persistent DOM, react to events to update the DOM, and serve the DOM to consumers. It depends on:

Reviewer: ```suggestion
At the core of Rojo lies `ServeSession`. As the name implies, it contains all of the components to keep a persistent DOM, react to events to update the DOM, and serve the DOM to consumers. It depends on:
```
Clunky and grammatically incorrect sentence structure

- `RojoTree` to represent the DOM;

> - `Project` to represent your root project file (e.g. `default.project.json`).

Reviewer: ```suggestion
- `Project` to represent your root project file (e.g. `default.project.json`);
```
Fix inconsistent line ending

## Overarching

Rojo is divided into several components, each layering on top of each other provide Rojo's functionality.

At the core of Rojo lies [`ServeSession`](#servesession). As the name implies, it contains all of the components to keep a persistent DOM, react to events to update the DOM, and serve the DOM to consumers.

Most of Rojo's uses are built upon `ServeSession`! For example, [`SourcemapCommand`](#sourcemapCommand) uses `ServeSession` to generate the DOM and read it to build the `sourcemap.json` file.

### The Serve Command

There are two main pieces in play when serving: the server and the Studio plugin.

The server runs a local [`LiveServer`](#liveserver) with access to your filesystem (whether it be via the terminal, the visual studio code extension, or a remote machine). It consumes a `ServeSession` and attaches a web server on top. The web server itself is very basic, consisting of around half a dozen endpoints. Generally, [`LiveServer`](#liveserver) acts as a middleman with the bulk of the work is performed by either the underlying `ServeSession` or the plugin. 

> To serve a project to a connecting plugin, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. After that, when something changes on the file system, the underlying `ServeSession` emits new patches. The web server has a `api/subscribe` endpoint where the plugin [long polls](https://en.wikipedia.org/wiki/Push_technology#Long_polling) to receive the patches from the server and apply them to the datamodel in Studio.

Reviewer 2: ```suggestion
To serve a project to a connecting plugin, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. After that, when something changes on the file system, the underlying `ServeSession` emits new patches. The web server has an endpoint the plugin [long polls](https://en.wikipedia.org/wiki/Push_technology#Long_polling) to receive the patches from the server and apply them to the datamodel in Studio.
```
We should document endpoints seperately

Because snapshots are designed to be translated into Instances anyway, this system is also used by the `build` command to turn a Rojo project into a complete file. The backend for serializing a snapshot into a file is provided by `rbx-dom`, which is a different project.

## Serve Command

> There are two main pieces in doing this: the server and the Studio Plugin.

Reviewer: ```suggestion
There are two main pieces in play when serving: the server and the Studio plugin.
```
Slightly clearer wording and consistent capitalization

> The server runs a local `LiveServer` on your computer (be it through a terminal, the visual studio extension, or a remote machine). It consumes a `ServeSession` and attaches a web server on top. The web server itself is very basic, consisting of around half a dozen endpoints. Generally, `LiveServer` acts as a middleman with the bulk of the work is performed by either the underlying `ServeSession` or the plugin. 

Reviewer: ```suggestion
The server runs a local `LiveServer` with access to your filesystem (whether it be via the terminal, the visual studio code extension, or a remote machine). It consumes a `ServeSession` and attaches a web server on top. The web server itself is very basic, consisting of around half a dozen endpoints. Generally, `LiveServer` acts as a middleman with the bulk of the work is performed by either the underlying `ServeSession` or the plugin. 
```
Use updated wording from previous review

> To serve a project to a connecting plugin, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. After that, when something changes on the file system, the underlying `ServeSession` emits new patches. The web server has a `api/subscribe` endpoint where the plugin [long polls](https://en.wikipedia.org/wiki/Push_technology#Long_polling) on to receive the patches from the server and apply them to the datamodel in Studio.

Reviewer: ```suggestion
To serve a project to a connecting plugin, the server gathers data on all of the files in that project, puts it into a nice format, and then sends it to the plugin. After that, when something changes on the file system, the underlying `ServeSession` emits new patches. The web server has a `api/subscribe` endpoint where the plugin [long polls](https://en.wikipedia.org/wiki/Push_technology#Long_polling) to receive the patches from the server and apply them to the datamodel in Studio.
```
Grammar correction

> When the plugin receives a patch from the server (whether it be the initial patch or any subsequent ones), the plugin reads through the patch and attempts to to apply the changes described by it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.

Reviewer: ```suggestion
When the plugin receives a patch it reads through the patch contents and attempts to to apply the changes described by it. Any sugar (the patch visualizer, as an example) happens on top of the patches received from the server.
```

This bit is not adding anything meaningful and is technically incorrect, so I think we should drop it. The sentence is more correct, more succinct, and just as informative.

To learn more, read about [`memofs` architecture](crates/memofs/ARCHITECTURE.md).

### LiveServer

> LiveServer underlies the [`serve` command](#the-serve-command) and provides the web server which clients (such as the plugin) can use to interface with [`ServeSession`](#servesession).

Reviewer 2: ```suggestion
LiveServer underlies the [`serve` command](#the-serve-command) and provides the web server which clients (such as the Studio plugin) can use to interface with a [`ServeSession`](#servesession).
```

> There's two parts to the API, the UI and the API clients use.

Reviewer 2: ```suggestion
The web server has two components: a UI and the API used by clients.
```

> The UI provides information about the current [`DOM`](#rojotree) has, including metadata. It also shows the project name, up-time, and version its Rojo is on.

Reviewer 2: ```suggestion
The UI provides information about the current project's [tree](#rojotree), including metadata. It also shows the project name, up-time, and version its Rojo is on.
```
This feels more accurate and avoids us having to explain the difference between the actual DOM and RojoTree later.

> The API provides a simple JSON protocol to interact with and receive changes from the underlying [`ServeSession`](#servesession). Checkout the [`api.rs` file under the web module](src/web/api.rs) to learn more.

Reviewer 2: This is fine for now, but we should eventually link this to the API documentation when it happens!

Author: Ah, I forgot that `doc.rust-lang.org` existed.

Author: Do you think we can use the [`web_api` crate doc page](https://docs.rs/rojo/latest/librojo/web_api/index.html) after brushing it up? If so, should we do it on this PR or a separate one?

Reviewer 2: I think yes, though we should also try to have documentation for the web endpoints that aren't Rust-user focused. If someone from e.g. `roblox-ts` (this is an example, I don't know of anyone who does) wanted to use our endpoints, they shouldn't have to read through Rust docs to know what to expect and how to use them.

### ServeSession

> The linchpin of Rojo. It contains all of the required components to serve a given project file.

Reviewer 2: ```suggestion
The `ServeSession` is the core of Rojo. It contains all of the required components to serve a given project file.
```

As much as I like the word linchpin, I feel like we should prioritize clarity here.

Generally, to serve means:

- Rojo maintains a DOM and exposes it to consumers;
- Rojo is able to accept events to cause changes to the DOM;
- Rojo is able to emit changes to the DOM to consumer.

It depends on:

> - `RojoTree` to represent the DOM;

Reviewer 2: ```suggestion
- [`RojoTree`](#rojotree) to represent the DOM;
```

- `Project` to represent your root project file (e.g. `default.project.json`);
- [`Vfs`](#vfs) to provide a filesystem and emit events on changes;
- [`ChangeProcessor`](#changeprocessor) to process filesystem events from `Vfs` and consequently update the DOM through the [snapshotting system](#the-snapshotting-system);

> It also provides an API for the higher level components so it can be used with the outside world.
> 
> - There is a `MessageQueue` of changes applied to the DOM.
> - There is a channel to send changes to the `ServeSession` and update the DOM.
> - And a `SessionId` to uniquely identify the `ServeSession`.
> 
> All of the public interfaces via CLI of Rojo are implemented using `ServeSession`.

Reviewer 2: ```suggestion
The primary interface for a `ServeSession` is the web server used by the [`serve` command](#servecommand). Additionally, several of Rojo's commands are implemented using `ServeSession`.
```

We should avoid getting too technical here. None of Rojo's existing commands actually interact directly with the `MessageQueue`, it's mostly an implementation detail.

Author: I see where you're coming from about getting too technical. I like the better summary you put it in. However, I think it would be weird omitting the other half of the `ServeSession` API. Especially because there wouldn't balance the first part of the internals and that most of Rojo's command do use to enable watching.

Reviewer 2: I suppose we could mention that it has a `MessageQueue`, though I'm inclined to just explain what that is rather than just saying "there is a `MessageQueue`"

Author: Alright, so I think I applied what was suggested. What's your thoughts on it?
